pkg/volume/csi/csi_attacher.go

```diff
@@ -64,88 +64,124 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
 	}
 
-	node := string(nodeName)
-	pvName := spec.PersistentVolume.GetName()
-	attachID := getAttachmentName(csiSource.VolumeHandle, csiSource.Driver, node)
+	klog.V(4).Info(log("attacher.Attach finished OK with VolumeAttachment object [%s]", createdAttchID))
 
-	attachment := &storage.VolumeAttachment{
-		ObjectMeta: meta.ObjectMeta{
-			Name: attachID,
-		},
-		Spec: storage.VolumeAttachmentSpec{
-			NodeName: node,
-			Attacher: csiSource.Driver,
-			Source: storage.VolumeAttachmentSource{
-				PersistentVolumeName: &pvName,
-			},
-		},
-		Status: storage.VolumeAttachmentStatus{Attached: false},
+	return createdAttchID, nil
+}
+
+// postVolumeAttachment posts the VolumeAttachment object to the server and waits for
+// the attachment to be marked attached.
+func (c *csiAttacher) postVolumeAttachment(driverName, volumeHandle string, attachment *storage.VolumeAttachment, attachmentTimeout time.Duration) (string, error) {
+
+	if attachment == nil {
+		return "", errors.New("VolumeAttachment nil")
 	}
 
-	_, err = c.k8s.StorageV1beta1().VolumeAttachments().Create(attachment)
+	klog.V(4).Info(log("attacher.postVolumeAttachment posting attachment [driver:%s, volumeHandle: %s, attachID:%s]", driverName, volumeHandle, attachment.Name))
+
+	createdAttachment, err := c.k8s.StorageV1beta1().VolumeAttachments().Create(attachment)
 	alreadyExist := false
 	if err != nil {
 		if !apierrs.IsAlreadyExists(err) {
-			klog.Error(log("attacher.Attach failed: %v", err))
+			klog.Error(log("attacher.postVolumeAttachment VolumeAttachments().Create() failed: %v", err))
 			return "", err
 		}
 		alreadyExist = true
 	}
 
 	if alreadyExist {
-		klog.V(4).Info(log("attachment [%v] for volume [%v] already exists (will not be recreated)", attachID, csiSource.VolumeHandle))
+		klog.V(4).Info(log("attachment [%v] for volume [%v] already exists (will not be recreated)", createdAttachment.Name, volumeHandle))
 	} else {
-		klog.V(4).Info(log("attachment [%v] for volume [%v] created successfully", attachID, csiSource.VolumeHandle))
+		klog.V(4).Info(log("attachment [%v] for volume [%v] created successfully", createdAttachment.Name, volumeHandle))
 	}
 
-	if _, err := c.waitForVolumeAttachment(csiSource.VolumeHandle, attachID, csiTimeout); err != nil {
+	attachID, err := c.waitForVolumeAttachment(driverName, volumeHandle, createdAttachment.Name, attachmentTimeout)
+	if err != nil {
+		klog.Error(log("attacher.postVolumeAttachment failed: %v", err))
 		return "", err
 	}
 
-	klog.V(4).Info(log("attacher.Attach finished OK with VolumeAttachment object [%s]", attachID))
-
 	return attachID, nil
 }
 
 func (c *csiAttacher) WaitForAttach(spec *volume.Spec, attachID string, pod *v1.Pod, timeout time.Duration) (string, error) {
-	source, err := getCSISourceFromSpec(spec)
+	klog.V(4).Info(log("attacher.WaitForAttach called [attachment.ID=%v]", attachID))
+	if spec == nil {
+		klog.Error(log("attacher.WaitForAttach missing volume.Spec"))
+		return "", errors.New("missing spec")
```

Comment:
I would prefer we don't log errors here again, and please put the better error message into the error that is returned, this will then get logged by the caller